feat(CDC) Snapshot consumer bootstrap state

fpacifici opened the pull request:
This provides a working implementation for the bootstrap state for the snapshot consumer.

This is the state the snapshot consumer starts into. When we starts a consumer of a dataset that is able to load snapshot these things happen in sequence:
1) we consume the cdc_control topic to know if there is a pending snapshot we need to catch up with 
2) if there is no snapshot we commit the cdc_control and start consuming the main cdc topic with the batching consumer
3) if there is a pending snapshot, we would start in paused state and keep consuming the control topic waiting for the snapshot ready message
4) if we find a message that indicates the snapshot has been loaded, we switch to the cdc consumer but ignore all the transactions up to the snapshot itself.

This PR provides the logic that consumes the control topic, addresses corner cases (like overlapping snapshots), and decides which state to starts into.

Builds on top of #419  and #437

fpacifici committed:
Introduce base for stateful consumer

fpacifici committed:
Bootstrap state

fpacifici committed:
Merge branch 'master' into feat/statefulConsumer

fpacifici committed:
Fix tests

fpacifici committed:
Refactor state machine

fpacifici committed:
Centralize the management of the state machine

fpacifici committed:
Merge branch 'master' into feat/statefulConsumer

fpacifici committed:
Merge branch 'feat/statefulConsumer' into feat/bootstrapState

fpacifici committed:
add the control protocol

fpacifici committed:
Protocol Schema

fpacifici committed:
Substate in bootstrap

fpacifici committed:
Split protocol out

fpacifici committed:
Add strict consumer

fpacifici committed:
Close the consumer at the end

fpacifici committed:
Fill in bootstrap state

fpacifici committed:
Add test

fpacifici committed:
Merge branch 'feat/bootstrapState' into feat/performBootstrap

fpacifici committed:
Fill in the bootstrap state

fpacifici committed:
Merge branch 'master' into feat/statefulConsumer

fpacifici committed:
Actually interpret the content of the control topic

fpacifici committed:
Give state a better typing

fpacifici committed:
Merge branch 'feat/statefulConsumer' into feat/bootstrapState

fpacifici committed:
Merge branch 'feat/bootstrapState' into feat/performBootstrap

fpacifici committed:
Some better typing and some additional comments

fpacifici committed:
Merge branch 'feat/statefulConsumer' into feat/bootstrapState

fpacifici committed:
Merge branch 'feat/bootstrapState' into feat/performBootstrap

fpacifici committed:
Add some comments

fpacifici committed:
Merge branch 'master' into feat/statefulConsumer

fpacifici committed:
review comments

fpacifici committed:
More refinements

fpacifici committed:
Merge branch 'feat/statefulConsumer' into feat/bootstrapState

fpacifici committed:
merge

fpacifici committed:
Refinements

fpacifici committed:
Merge branch 'feat/bootstrapState' into feat/performBootstrap

fpacifici committed:
Refinements

fpacifici committed:
Apparently I cannot change the english language to make Builer a legit word

fpacifici committed:
Merge branch 'feat/statefulConsumer' into feat/bootstrapState

fpacifici committed:
Merge branch 'feat/bootstrapState' into feat/performBootstrap

fpacifici committed:
Nits

fpacifici committed:
FIx tests

fpacifici committed:
Remove TStateType

fpacifici committed:
Fi tests

fpacifici committed:
Merge branch 'feat/statefulConsumer' into feat/bootstrapState

fpacifici committed:
Merge branch 'feat/bootstrapState' into feat/performBootstrap

fpacifici committed:
Fix tests

fpacifici committed:
Fix /query

fpacifici committed:
Merge branch 'master' into feat/bootstrapState

fpacifici committed:
Review changes

fpacifici committed:
Merge branch 'master' into feat/bootstrapState

fpacifici committed:
Merge branch 'master' into feat/bootstrapState

fpacifici committed:
Enforce one partition only

fpacifici committed:
Prevent partition reset

fpacifici committed:
Merge branch 'feat/bootstrapState' into feat/performBootstrap

fpacifici committed:
Fix test

fpacifici committed:
Fix shutdown

fpacifici committed:
Merge branch 'feat/bootstrapState' into feat/performBootstrap

fpacifici committed:
Fix message

tkaemming commented on snuba/stateful_consumer/control_protocol.py:
```diff
@@ -0,0 +1,172 @@
+                    "type": "string"
+                }
+            },
+            "required": ["event", "snapshot-id"],
```
Why is "event" part of this `required` list? It's not part of the property mapping, and seems like it's redundant with the definitions below?

fpacifici replied:
I think I just forgot it there. It took me a while to get this schema right, Likely I missed this one.

tkaemming commented on snuba/stateful_consumer/control_protocol.py:
```diff
@@ -0,0 +1,172 @@
+                {
+                    "properties": {
+                        "event": {"const": "snapshot-loaded"},
+                        "datasets": {
```
Why is this plural? It looks like only one can be included. Why isn't this itself just a reference to the definition directly, rather than through an object type and `additionalProperties` below?

fpacifici replied:
When I designed the protocol (before doing any test around groupedmessage dataset) I was expecting a consumer to be able to reload multiple datasets at a time. 
This is not likely to be the case, but in the PR above, I mentioned I want to entirely remove this field.

tkaemming commented on snuba/stateful_consumer/control_protocol.py:
```diff
@@ -0,0 +1,172 @@
+                                        {"type": "number"}
+                                    ],
+                                },
+                            }
```
Shouldn't all of `xmin`/`xmax`/`xip-list` be required?

fpacifici replied:
just forgot

tkaemming commented on snuba/stateful_consumer/states/bootstrap.py:
```diff
@@ -1,25 +1,101 @@
+        if msg.product != settings.SNAPSHOT_LOAD_PRODUCT:
+            return
+        if self.__active_snapshot_msg:
+            if isinstance(self.__active_snapshot_msg, SnapshotInit):
```
When would this not be true? What type is `__active_snapshot_msg`?

Later: OK. I saw this assigned again below. Ideally the design of this class would be explained in the docstring or a comment or something, rather than requiring the reader to read the class first and figure it out for themself from the implementation

fpacifici replied:
Ok, I will beef up the docstring

tkaemming commented on snuba/stateful_consumer/states/bootstrap.py:
```diff
@@ -1,25 +1,101 @@
+                "Duplicate Snapshot init: %r",
+                msg.id,
+            )
+        self.__processed_snapshots.add(msg.id)
```
Is "processed" really an accurate name for this? "Processed" seems to imply the entire snapshot process has been completed, and that doesn't seem to be the case here.

tkaemming commented on snuba/stateful_consumer/states/bootstrap.py:
```diff
@@ -38,10 +114,32 @@ def __init__(self,
+        value = json.loads(message.value())
+        parsed_message = parse_control_message(value)
+
+        current_snapshot = self.__recovery_state.get_active_snapshot_msg()
```
This assignment doesn't have anything to do with the code immediately following it, it should be moved down closer to the usage unless there's some reason it needs to be here (if there is, it should be commented.)

fpacifici replied:
The lines below willchange the result of get_active_snapshot_msg, so this has to be before. It will be compared with new_snap later

tkaemming commented on snuba/stateful_consumer/states/bootstrap.py:
```diff
@@ -38,10 +114,32 @@ def __init__(self,
+            logger.debug("Committing offset %r ", message.offset())
+            return CommitDecision.COMMIT_THIS
+        elif current_snapshot is None or new_snap.id != current_snapshot.id:
+            logger.debug("Committing previous offset to %r ", message.offset())
```
```suggestion
            logger.debug("Committing previous offset to %r", message.offset())
```

tkaemming commented on snuba/stateful_consumer/states/bootstrap.py:
```diff
@@ -38,10 +114,32 @@ def __init__(self,
+
+        new_snap = self.__recovery_state.get_active_snapshot_msg()
+        if new_snap is None:
+            logger.debug("Committing offset %r ", message.offset())
```
```suggestion
            logger.debug("Committing offset %r", message.offset())
```

tkaemming reviewed (changes requested):
Kind of struggling to see how the pieces fit together while reading this. I think elaborating on the design in the docstring or comments would make reviewing this now and understanding it in the future a lot more approachable.

fpacifici commented on snuba/consumers/strict_consumer.py:
```diff
@@ -142,7 +142,7 @@ def run(self) -> None:
                         offset=prev_watermark,
                     )
-                    self.__consumer.commit(commit_pos, asynchronous=False)
+                    self.__consumer.commit(offsets=[commit_pos], asynchronous=False)
```
It was supposed to be part of the previous PR, but none was using this code there.

fpacifici commented on snuba/stateful_consumer/states/bootstrap.py:
```diff
@@ -1,25 +1,101 @@
+        if self.__active_snapshot_msg:
+            if isinstance(self.__active_snapshot_msg, SnapshotInit):
+                logger.error(
+                    "Overlapping snapshots. Ignoring. Running %r. Init received %r.",
```
will clarify

fpacifici reviewed (commented):
I will add more details about the design in the RecoveryState docstring. Is that what you are referring to specifically . or is there something else that is not clear ?

fpacifici committed:
Merge branch 'master' into feat/performBootstrap

fpacifici committed:
Address comments

fpacifici committed:
Clarify RecoveryState

fpacifici committed:
Move datasets into tables

fpacifici commented:
Updates:
- added more documentation to the bootstrap state
- changed the way RecoveryState keeps track of the state itself. This should be more understandable
- Removed datasets from the snapshot-loaded message and added tables to snapshot-init so that the bootstrap state can ignore the tables it is not interested into.

tkaemming commented on snuba/stateful_consumer/states/bootstrap.py:
```diff
@@ -1,25 +1,178 @@
+        if self.__active_snapshot_id and not self.__loaded_snapshot_transactions:
+            logger.error(
+                "A new snapshot-init message was found before a previous one "
+                " was completed. Ignoring. Running %r. Init received %r.",
```
"running" here still strikes me as odd, since it doesn't change the active snapshot. Would it make more sense to word this as `Ignoring %r, and continuing with %r` instead?

It also seems a bit weird to use `error` when this isn't logging an exception, kind of feels like this should be a `warning` instead.

tkaemming commented on snuba/stateful_consumer/states/bootstrap.py:
```diff
@@ -1,25 +1,178 @@
+            return self.__commit_if_no_active_snapshot()
+        if self.__active_snapshot_id != msg.id:
+            logger.warning(
+                "Aborting a snapshot that is not active. Active %r, Abort %r",
```
I think here (and in `process_snapshot_below` it might make sense to reword this as noted above (if you agree with my interpretation of what this is saying.)

tkaemming reviewed (approved):
This is a lot easier to understand now, thanks!

There's some minor things inline (mostly typos or wording changes and a couple style things) but nothing serious or blocking.

fpacifici committed:
Address comments

fpacifici committed:
Remove context file

tkaemming commented on snuba/stateful_consumer/control_protocol.py:
```diff
@@ -0,0 +1,162 @@
+            "allOf": [
+                {"$ref": "#/definitions/base"},
+                {
+                    "properties": {
```
Same as here: https://github.com/getsentry/snuba/pull/441/files/090d0499a517fc2cffff244c9a3be69f51f22dfe#diff-b4875ea4684246d1a462c8d94f40ec6dR24
